https://bugs.gentoo.org/327347

fix from upstream for display of uncompressed non-English manpages

--- man-db-2.5.7/lib/pipeline.c
+++ man-db-2.5.7/lib/pipeline.c
@@ -329,6 +329,25 @@
 	return cmd;
 }
 
+static void passthrough (void *data ATTRIBUTE_UNUSED)
+{
+	for (;;) {
+		char buffer[4096];
+		int r = read (STDIN_FILENO, buffer, 4096);
+		if (r <= 0)
+			break;
+		if (fwrite (buffer, 1, (size_t) r, stdout) < (size_t) r)
+			break;
+	}
+
+	return;
+}
+
+command *command_new_passthrough (void)
+{
+	return command_new_function ("cat", &passthrough, NULL, NULL);
+}
+
 command *command_dup (command *cmd)
 {
 	command *newcmd = XMALLOC (command);
@@ -831,20 +850,6 @@
 	return p;
 }
 
-static void passthrough (void *data ATTRIBUTE_UNUSED)
-{
-	for (;;) {
-		char buffer[4096];
-		int r = read (STDIN_FILENO, buffer, 4096);
-		if (r <= 0)
-			break;
-		if (fwrite (buffer, 1, (size_t) r, stdout) < (size_t) r)
-			break;
-	}
-
-	return;
-}
-
 void pipeline_connect (pipeline *source, pipeline *sink, ...)
 {
 	va_list argv;
@@ -876,11 +881,8 @@
 		 * because it has nowhere to send output. Until this is
 		 * fixed, this kludge is necessary.
 		 */
-		if (arg->ncommands == 0) {
-			command *cmd = command_new_function
-				("cat", &passthrough, NULL, NULL);
-			pipeline_command (arg, cmd);
-		}
+		if (arg->ncommands == 0)
+			pipeline_command (arg, command_new_passthrough ());
 	}
 	va_end (argv);
 }
--- man-db-2.5.7/lib/pipeline.h
+++ man-db-2.5.7/lib/pipeline.h
@@ -172,6 +172,9 @@
  */
 command *command_new_sequence (const char *name, ...) ATTRIBUTE_SENTINEL;
 
+/* Return a new command that just passes data from its input to its output. */
+command *command_new_passthrough (void);
+
 /* Return a duplicate of a command. */
 command *command_dup (command *cmd);
 
--- man-db-2.5.7/NEWS
+++ man-db-2.5.7/NEWS
@@ -1,3 +1,15 @@
+man-db 2.5.8
+============
+
+Major changes since man-db 2.5.7:
+
+	Fixes:
+	------
+
+	o Fix assertion failure on 'man -l' with an uncompressed page and
+	  any of --no-hyphenation, --no-justification, or a non-English
+	  page.
+
 man-db 2.5.7 (16 February 2010)
 ===============================
 
--- man-db-2.5.7/src/man.c
+++ man-db-2.5.7/src/man.c
@@ -2390,9 +2390,16 @@
 #endif /* TROFF_IS_GROFF */
 
 		if (seq->u.sequence.ncommands) {
-			assert (decomp->ncommands == 1);
-			command_sequence_command (seq, decomp->commands[0]);
-			decomp->commands[0] = seq;
+			assert (decomp->ncommands <= 1);
+			if (decomp->ncommands) {
+				command_sequence_command
+					(seq, decomp->commands[0]);
+				decomp->commands[0] = seq;
+			} else {
+				command_sequence_command
+					(seq, command_new_passthrough ());
+				pipeline_command (decomp, seq);
+			}
 		} else
 			command_free (seq);
 	}

